### PR TITLE
Fix cacheWatcher leak when time jump to the future and jump back

### PR DESCRIPTION
**Pull request by chenyw1990**
What type of PR is this?
/kind bug

What this PR does / why we need it:
Fix cacheWatcher leak when time jump to the future and jump back

Which issue(s) this PR fixes:
#96259

```release-note
Fix memory leak in kube-apiserver when underlying time goes forth and back.
```

### CODE REVIEW

**Comment by fedebongio** [comment]
/triage accepted
/assign @wojtek-t 
(thank you!)

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 1]
```diff
@@ -189,15 +189,44 @@ func (t *watcherBookmarkTimeBuckets) addWatcher(w *cacheWatcher) bool {
...
+}
+
 func (t *watcherBookmarkTimeBuckets) popExpiredWatchers() [][]*cacheWatcher {
 	currentBucketID := t.clock.Now().Unix()
 	// There should be one or two elements in almost all cases
 	expiredWatchers := make([][]*cacheWatcher, 0, 2)
 	t.lock.Lock()
 	defer t.lock.Unlock()
+
+	if t.startBucketID > currentBucketID {
```
Comment (gongguan) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:226]: A naive question: should buckets between currentBucketID and t.startBucketID be handled?
By the way, this worth a comment.

**Comment by wojtek-t** [comment]
/hold

I would like to first understand how exactly that may happen to ensure that fix is what we want - pending answer on: https://github.com/kubernetes/kubernetes/issues/96259#issuecomment-723572297

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 1 above)
Comment (chenyw1990) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:226]: If we expire buckets between currentBucketID and t.startBucketID when t.startBucketID is bigger than currentBucketID, I think some bookmark pushes maybe failed.
In this modification, the cacheWatcher reference in watchersBuckets will be actively cleaned up when the cacheWatcher is closed. I think this is better.

**Comment by liggitt** [comment]
Rather than add a mechanism like this, can we modify the existing time calculations to make use of the [golang monotonic clock](https://golang.org/pkg/time/#hdr-Monotonic_Clocks), which is not affected by forward/backward changes in wall clock time?

**Comment by chenyw1990** [comment]
@liggitt @wojtek-t @gongguan 
I have refactored the code using golang's monotonic clock feature and have passed the test in our product.
review please.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 2]
```diff
@@ -181,7 +184,7 @@ func (t *watcherBookmarkTimeBuckets) addWatcher(w *cacheWatcher) bool {
 	if !ok {
 		return false
 	}
-	bucketID := nextTime.Unix()
+	bucketID := nextTime.Sub(t.createTime) / time.Second
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:187]: Can we add a unit test to validate that we're no longer leaking watchers?

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go [hunk 3]
```diff
@@ -129,12 +129,6 @@ func (s *Scheme) nameFunc(t reflect.Type) string {
 	return gvks[0].Kind
 }
 
-// fromScope gets the input version, desired output version, and desired Scheme
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:132]: Please move this commit to a separate PR.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 2 above)
Comment (chenyw1990) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:187]: Yeah, I think that's a good idea. I'll add a unit test.

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go (hunk 3 above)
Comment (chenyw1990) [reply staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: i have moved it.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 4]
```diff
@@ -159,16 +159,18 @@ func (i *indexedWatchers) terminateAll(objectType reflect.Type, done func(*cache
 // if you set fire time at X, you can get the bookmark within (X-1,X+1) period.
 type watcherBookmarkTimeBuckets struct {
 	lock              sync.Mutex
-	watchersBuckets   map[int64][]*cacheWatcher
-	startBucketID     int64
+	watchersBuckets   map[time.Duration][]*cacheWatcher
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:162]: suggest leaving the key as int64 to avoid confusion about granularity of the map keys

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 5]
```diff
@@ -159,16 +159,18 @@ func (i *indexedWatchers) terminateAll(objectType reflect.Type, done func(*cache
...
+	createTime        time.Time
+	startBucketID     time.Duration
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:164]: same comment here, leave as int64

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 2 above)
Comment (chenyw1990) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:186]: I think modifying the system time in unit test is not a good idea, there is currently no good idea to test this scenario

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:162]: since we're only storing values on second granularity, I think making the key a duration causes 1,000,000,000 empty lookups as we're iterating from one second to the next

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 2 above)
Comment (chenyw1990) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:186]: I think modifying the system time in unit test is not a good idea, there is currently no good idea to test this scenario

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:162]: also add a comment that the watchersBucket key is the number of seconds since `createTime`

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 6]
```diff
@@ -181,7 +184,7 @@ func (t *watcherBookmarkTimeBuckets) addWatcher(w *cacheWatcher) bool {
 	if !ok {
 		return false
 	}
-	bucketID := nextTime.Unix()
+	bucketID := int64(nextTime.Sub(t.createTime) / time.Second)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:187]: Going back to my original comment about adding a test - can't you use FakeClock to test this?

Comment (chenyw1990) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:187]: The essence of this problem is that the display clock changes, but the mononotonic clock does not change.
The FakeClock time can be changed using either of the following methods: SetTime and Step, both methods change FakeClock.time. The display clock and monotonic clock change simultaneously. Therefore, the problem cannot be simulated.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:187]: OK - that sounds reasonable.

@liggitt - do you have any suggestions what else can we do here? Or can we go with as is?

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:187]: I also don't see a way to exercise the bug with the new code, so I'd be happy with this fix as is

**Comment by wojtek-t** [comment]
/lgtm
/approve
/hold cancel

/milestone v1.20
[Adding milestone as this is bug fix - we will be cherrypicking back once this merges]
